File: src/services/volte.ts
/**
 * Voice-over-Cellular / VoLTE (ap2003) Service Handler
 *
 * Builds the ApplicationConfig for VoLTE and VoNR entitlements.
 */

import type { ApplicationConfig } from '../protocol/responseTypes.js';
import { EntitlementStatus } from '../protocol/statusCodes.js';

interface VoLTEConfigData {
  addresses?: Array<{ addrType: string; addr: string }>;
  volteEntitled?: string;
  vonrEntitled?: string;
}

const DEFAULT_ADDRESSES = [
  { addrType: '1', addr: 'pcscf.operator.com' },
];

export function buildVoLTEConfig(
  status: number,
  provStatus: number,
  tcStatus: number,
  configData?: unknown,
): ApplicationConfig {
  const config = (configData ?? {}) as VoLTEConfigData;

  const result: ApplicationConfig = {
    appId: 'ap2003',
    entitlementStatus: status,
    tcStatus,
    provStatus,
    extraParams: {},
  };

  if (status === EntitlementStatus.ENABLED) {
    result.addrStatus = 1;
    result.addresses = config.addresses ?? DEFAULT_ADDRESSES;
    result.extraParams!.VoLTE_Entitled = config.volteEntitled ?? '1';
    result.extraParams!.VoNR_Entitled = config.vonrEntitled ?? '1';
  } else {
    result.addrStatus = 0;
    result.extraParams!.VoLTE_Entitled = '0';
    result.extraParams!.VoNR_Entitled = '0';
  }

  return result;
}

File: internal/services/volte.go
package services

import (
	"github.com/hkassaei/entitlements-as-a-service-poc-go/internal/protocol"
)

// VoLTEConfigData holds configData for ap2003.
type VoLTEConfigData struct {
	Addresses     []protocol.AddressConfig `json:"addresses,omitempty"`
	VoLTEEntitled string                   `json:"volteEntitled,omitempty"`
	VoNREntitled  string                   `json:"vonrEntitled,omitempty"`
}

var defaultVoLTEAddresses = []protocol.AddressConfig{
	{AddrType: "1", Addr: "pcscf.operator.com"},
}

// BuildVoLTEConfig builds the ApplicationConfig for VoLTE (ap2003).
func BuildVoLTEConfig(status, provStatus, tcStatus int, configData *VoLTEConfigData) *protocol.ApplicationConfig {
	if configData == nil {
		configData = &VoLTEConfigData{}
	}

	result := &protocol.ApplicationConfig{
		AppID:             "ap2003",
		EntitlementStatus: status,
		TcStatus:          protocol.IntPtr(tcStatus),
		ProvStatus:        protocol.IntPtr(provStatus),
		ExtraParams:       map[string]string{},
	}

	if status == protocol.EntitlementStatusEnabled {
		result.AddrStatus = protocol.IntPtr(1)
		if len(configData.Addresses) > 0 {
			result.Addresses = configData.Addresses
		} else {
			result.Addresses = defaultVoLTEAddresses
		}
		volte := configData.VoLTEEntitled
		if volte == "" {
			volte = "1"
		}
		vonr := configData.VoNREntitled
		if vonr == "" {
			vonr = "1"
		}
		result.ExtraParams["VoLTE_Entitled"] = volte
		result.ExtraParams["VoNR_Entitled"] = vonr
	} else {
		result.AddrStatus = protocol.IntPtr(0)
		result.ExtraParams["VoLTE_Entitled"] = "0"
		result.ExtraParams["VoNR_Entitled"] = "0"
	}

	return result
}
